lib/pingurl.js

```diff
@@ -0,0 +1,44 @@
+'use strict';
+
+const url = require('url');
+const http = require('http');
+const https = require('https');
+
+/**
+ * Attempt to GET a url, with retries/error handling
+ */
+exports.ping = (pingUrl, timeout) => {
+  if (timeout === undefined) {
+    timeout = 2000;
+  }
+  return new Promise((resolve, reject) => {
+    httpGet(pingUrl, timeout, resolve, reject, 3);
+  });
+}
+
+/**
+ * Retry handler
+ */
+function httpGet(pingUrl, timeout, resolve, reject, retries) {
+  let parsed = url.parse(pingUrl);
+  let getter = parsed.protocol === 'https:' ? https : http;
+  if (!parsed.host) {
+    return reject(new Error(`Invalid ping url: ${pingUrl}`));
+  }
+
+  // get the url, retrying timeouts
+  getter.get(pingUrl, (res) => {
+    if (res.statusCode === 502 && retries > 1) {
+      httpGet(pingUrl, timeout, resolve, reject, retries - 1);
```

Comment:
retry immediately? wonder if it should sleep or something...

Reply:
Good idea!  Related - any opinions/ideas on how long the individual timeouts should be on this pixel-tracker?  (Or on pingbacks in general?)

Currently, 2000ms * 3-retries = need at least a 6s timeout on the lambda in general.  Current lambdas timeout at 30s (because BQ+retries can be slower) ... so plenty of overhead if we want a much longer wait.

Reply:
And now that I'm thinking about it ...

ONLY 502s are getting retried right now, a maximum of 3 times.  Timeouts don't get retried at all.  I went with this for now because (1) we don't generally see timeouts in DT, and (2) I'd rather under-report impressions than over-report.

If (1) changes, we'll have to update this to also catch timeouts.